File: CPP/2-Structurals/Proxy/Proxy.cpp
#include <iostream>

class RealImage
{
    int _id;

public:
    RealImage(int i)
    {
        _id = i;
        std::cout << "   ctor: " << _id << std::endl;
    }

    ~RealImage()
    {
        std::cout << "   dtor: " << _id << std::endl;
    }

    void draw()
    {
        std::cout << "   drawing the image " << _id << std::endl;
    }
};

// Design an "extra level of indirection" proxy class
class ImageProxy
{
    // The proxy class holds a pointer to the real class
    RealImage *_the_real_image;
    int _id;
    static int _next;

public:
    ImageProxy()
    {
        _id = _next++;
        // Initialized to null
        _the_real_image = NULL;
    }

    ~ImageProxy()
    {
        if (_the_real_image != NULL)
        {
            delete _the_real_image;
        }
    }

    void draw()
    {
        // When a request comes in, the real object is
        //    created "on first use"
        if (NULL == _the_real_image)
        {
            _the_real_image = new RealImage(_id);
        }
        // The request is always delegated
        _the_real_image->draw();
    }
};

// Initialize the proxy’s counter
int ImageProxy::_next = 1;

int main()
{
    ImageProxy images[5];

    for (int i = 0; i < 5; ++i)
    {
        images[i].draw();
    }

    return 0;
}

//-----------------------------------------------------------
/*
Output:

ctor: 1
drawing image 1
ctor: 2
drawing image 2
ctor: 3
drawing image 3
ctor: 4
drawing image 4
ctor: 5
drawing image 5
dtor: 5
dtor: 4
dtor: 3
dtor: 2
dtor: 1

*/

File: Python/2-Structurals/Proxy/Proxy.py
#!/usr/bin/env python3


class RealImage:
    def __init__(self, idx):
        self.image_id = idx
        print("Build RealImage with ID: %d" % idx)

    def draw(self):
        print("drawing the image %d" % self.image_id)


# Design an "extra level of indirection" proxy class
class ImageProxy:
    _next_id = 0

    def __init__(self):
        self._the_real_image = None
        ImageProxy._next_id += 1
        self.idx = ImageProxy._next_id

    def draw(self):
        # When a request comes in, the real object is created "on first use"
        if self._the_real_image is None:
            self._the_real_image = RealImage(self.idx)

        self._the_real_image.draw()


def main():
    print("Init proxies...")
    images = [ImageProxy() for _ in range(0, 5)]

    print("\nDrawing images... \n")
    for image in images:
        image.draw()


if __name__ == "__main__":
    main()


"""
Output:

Init proxies...

Drawing images... 

Build RealImage with ID: 1
drawing the image 1
Build RealImage with ID: 2
drawing the image 2
Build RealImage with ID: 3
drawing the image 3
Build RealImage with ID: 4
drawing the image 4
Build RealImage with ID: 5
drawing the image 5
"""
